# Fashion MNIST

## MNIST 패션 이미지를 CNN을 이용하여 분류하세요. CNN을 이용한 분류 결과를 MLP와 비교해보세요.

In [1]:
from tensorflow import keras
from keras import models 
from keras import layers
import numpy as np
import matplotlib.pyplot as plt
from keras.utils import to_categorical
from keras.datasets import fashion_mnist

### Multi-Layer Perceptron (MLP)

In [2]:
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()
y_train=to_categorical(y_train)
y_test=to_categorical(y_test)

# preprocessing
X_train = X_train.reshape((60000, 28*28))
X_train = X_train/255

X_test = X_test.reshape((10000, 28*28))
X_test = X_test/255

net = models.Sequential()
net.add(layers.Dense(512, activation='relu', input_shape=(28*28,))) 
net.add(layers.Dense(512, activation='relu')) 
net.add(layers.Dropout(0.5))
net.add(layers.Dense(512, activation='relu')) 
net.add(layers.Dropout(0.5))
net.add(layers.Dense(10, activation='softmax')) 
net.compile(optimizer='nadam',loss='categorical_crossentropy',metrics=['acc'])

In [3]:
X_val = X_train[:10000]
partial_X_train = X_train[10000:]

y_val = y_train[:10000]
partial_y_train = y_train[10000:]

In [4]:
val= net.fit(partial_X_train,partial_y_train,epochs=10, batch_size=64, validation_data=(X_val, y_val))

Epoch 1/10
782/782 [==============================] - 30s 38ms/step - loss: 0.5615 - acc: 0.7983 - val_loss: 0.4285 - val_acc: 0.8507
Epoch 2/10
782/782 [==============================] - 28s 36ms/step - loss: 0.3984 - acc: 0.8571 - val_loss: 0.4046 - val_acc: 0.8564
Epoch 3/10
782/782 [==============================] - 28s 36ms/step - loss: 0.3567 - acc: 0.8701 - val_loss: 0.4087 - val_acc: 0.8521
Epoch 4/10
782/782 [==============================] - 28s 36ms/step - loss: 0.3304 - acc: 0.8805 - val_loss: 0.3344 - val_acc: 0.8804
Epoch 5/10
782/782 [==============================] - 28s 36ms/step - loss: 0.3123 - acc: 0.8867 - val_loss: 0.3765 - val_acc: 0.8626
Epoch 6/10
782/782 [==============================] - 28s 36ms/step - loss: 0.3012 - acc: 0.8909 - val_loss: 0.3169 - val_acc: 0.8843
Epoch 7/10
782/782 [==============================] - 28s 36ms/step - loss: 0.2906 - acc: 0.8923 - val_loss: 0.3215 - val_acc: 0.8785
Epoch 8/10
782/782 [==============================] - 28s 36ms

In [5]:
test_loss, test_acc = net.evaluate(X_test,y_test)

313/313 [==============================] - 2s 6ms/step - loss: 0.3732 - acc: 0.8729


# CNN

In [11]:
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

# preprocessing
X_train = X_train.reshape((60000, 28, 28, 1))
X_train = X_train/255

X_test = X_test.reshape((10000, 28, 28, 1))
X_test = X_test/255

y_train=to_categorical(y_train)
y_test=to_categorical(y_test)

In [12]:
net = models.Sequential()

net.add(layers.Conv2D(32,(3,3), activation='relu', input_shape=(28,28,1)))
net.add(layers.MaxPooling2D((2,2)))
net.add(layers.Conv2D(64,(3,3), activation='relu')) 
net.add(layers.MaxPooling2D((2,2)))
net.add(layers.Conv2D(64,(3,3), activation='relu'))

net.add(layers.Flatten())

net.add(layers.Dense(64, activation='relu')) 
net.add(layers.Dense(10, activation='softmax')) 

net.compile(optimizer='nadam',loss='categorical_crossentropy',metrics=['acc'])

In [13]:
net.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_3 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_1 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_6 (Dense)              (None, 64)               

In [14]:
net.fit(X_train, y_train, epochs=5, batch_size=64)

Epoch 1/5
938/938 [==============================] - 57s 61ms/step - loss: 0.5300 - acc: 0.8041
Epoch 2/5
938/938 [==============================] - 56s 60ms/step - loss: 0.3319 - acc: 0.8793
Epoch 3/5
938/938 [==============================] - 63s 67ms/step - loss: 0.2837 - acc: 0.8965
Epoch 4/5
938/938 [==============================] - 75s 80ms/step - loss: 0.2544 - acc: 0.9080
Epoch 5/5
938/938 [==============================] - 69s 74ms/step - loss: 0.2311 - acc: 0.9162


In [15]:
test_loss, test_acc = net.evaluate(X_test,y_test)

313/313 [==============================] - 3s 11ms/step - loss: 0.2762 - acc: 0.9005


## MLP 결과 : 0.8729의 정확도
## CNN 결과 : 0.9005의 정확도

 이미지 인식에 있어서 MLP는 모든 입력이 위치와 상관없이 동일한 수준의 중요도를 갖고, 이를 이용해 neural network를 구성하면 파라미터의 크기가 엄청나게 커지는 문제가 생긴다. 이 문제를 해결하기 위해 만들어진 것이 CNN이다.
 CNN은 max-pooling 방식의 sub-sampling 과정을 거치는데, 이 과정에서 각각 가장 큰 자극을 선택한다. 이 과정이 데이터의 특징과 패턴을 파악하는 것이고, 이미지를 분류하는 데 있어 이미지 전체에 대한 특징을 얻어내기 때문에 인식능력이 좋아진다. 
 따라서 결과를 비교해 보았을 때 CNN으로 분류한다면 성능이 높아지는 것을 볼 수 있다.